src/react/utils.js

```diff
@@ -58,6 +60,95 @@ export function valueIsClassComponent(realm: Realm, value: Value) {
+function canHoistObject(realm: Realm, object: ObjectValue, hoistedScope?: Generator): boolean {
+  if (isReactElement(object)) {
+    return canHoistReactElement(realm, object);
+  }
+  for (let [propName] of object.properties) {
+    let prop = Get(realm, object, propName);
+    if (prop instanceof AbstractValue) {
+      return false;
+    } else if (prop instanceof FunctionValue && prop.parent !== hoistedScope) {
+      return false;
+    } else if (prop instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (prop instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, prop, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+function canHoistArray(realm: Realm, array: ArrayValue, hoistedScope?: Generator): boolean {
+  let lengthValue = Get(realm, array, "length");
+  invariant(lengthValue instanceof NumberValue);
+  let length = lengthValue.value;
+  for (let i = 0; i < length; i++) {
+    let element = Get(realm, array, "" + i);
+
+    if (element instanceof AbstractValue) {
+      return false;
+    } else if (element instanceof ArrayValue) {
+      let isHoistable = canHoistArray(realm, element, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    } else if (element instanceof FunctionValue && element.parent !== hoistedScope) {
+      return false;
+    } else if (element instanceof ObjectValue) {
+      let isHoistable = canHoistObject(realm, element, hoistedScope);
+
+      if (!isHoistable) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+export function canHoistReactElement(realm: Realm, reactElement: ObjectValue, hoistedScope?: Generator): boolean {
+  if (reactElement.$CanHoist) {
+    return true;
+  }
+  // we host components <div />, <span /> etc
+  // we also allow functions where their parent scope matches the hoisted scope
+  let type = Get(realm, reactElement, "type");
+  if (!(type instanceof StringValue || (type instanceof FunctionValue && type.parent === hoistedScope))) {
+    return false;
+  }
+  // you can only hoist null or function values where the scope matches the hoisted scope
+  let ref = Get(realm, reactElement, "ref");
+  if (!(ref === realm.intrinsics.null || (ref instanceof FunctionValue && ref.parent === hoistedScope))) {
+    return false;
+  }
+  // keys should also be string or null
+  let key = Get(realm, reactElement, "key");
+  if (!(key instanceof StringValue || key === realm.intrinsics.null)) {
+    return false;
+  }
+  let props = Get(realm, reactElement, "props");
+  invariant(props instanceof ObjectValue);
+  let isHoistable = canHoistObject(realm, props, hoistedScope);
+  if (!isHoistable) {
+    return false;
+  }
+  // cache the value so we can avoid doing this work again
+  reactElement.$CanHoist = true;
```

Comment:
You could use a Set or WeakSet for this instead so that we don't have to add another field to ObjectValue that is already so bloated, and has complexities like the merging of values in different values and changes over time. Currently the layering between the interpreter and the serialization step is pretty clean. Ideally the additional functions/reconciliation steps should also be cleanly separated from the serialization. This can start mattering more as we're considering new serialization formats such as byte codes. With bundle splitting you may also need to serialize this same value into multiple serialized outputs and then it's cleaner to have this concern exist once per serializer potentially.

Is there precedence for other serialization related caching using additional fields?

That said, nothing else needs something like this because it is added to the set that tracks which values are already visited. Can this just use that mechanism to bail out early?